threestudio/models/guidance/zero123_guidance.py

```diff
@@ -314,11 +316,155 @@ def __call__(
+
+        return guidance_out, guidance_eval_out
+
+    @torch.cuda.amp.autocast(enabled=False)
+    @torch.no_grad()
+    def guidance_eval(self, cond, t, latents, noise_pred):
+        # self.scheduler.set_timesteps(self.num_train_timesteps)
+        # index = torch.where(
+        #     self.scheduler.timesteps.reshape(-1, 1).repeat(len(t), 1) == t.cpu()
+        # )[0][: len(t)]
+        # use only 50 timesteps, and find nearest of those to t
+        self.scheduler.set_timesteps(50)
+        index = torch.min(
+            self.scheduler.timesteps.reshape(1, -1).repeat(2, 1)
+            > t.reshape(-1, 1).cpu(),
+            dim=1,
+        )[1]
+        t = self.scheduler.timesteps[index]
+
+        imgs_noisy = self.decode_latents(latents).permute(0, 2, 3, 1)
+
+        # get prev latent
+        latents_new = []
+        for b in range(len(t)):
+            latents_new.append(
+                self.scheduler.step(
+                    noise_pred[b : b + 1], t[b], latents[b : b + 1], eta=1
+                )["prev_sample"]
+            )
+        latents_new = torch.cat(latents_new)
+        imgs_1step = self.decode_latents(latents_new).permute(0, 2, 3, 1)
+
+        latents_final = []
+        # for i, time in enumerate(index):
+        for b, i in enumerate(index):
+            latents = latents_new[b : b + 1]
+            c = {
+                "c_crossattn": [cond["c_crossattn"][0][b * 2 : b * 2 + 2]],
+                "c_concat": [cond["c_concat"][0][b * 2 : b * 2 + 2]],
+            }
+            for t in tqdm(self.scheduler.timesteps[i + 1 :], leave=False):
+                # pred noise
+                x_in = torch.cat([latents] * 2)
+                t_in = torch.cat([t.reshape(1)] * 2).to(self.device)
+                noise_pred = self.model.apply_model(x_in, t_in, c)
+                # perform guidance
+                noise_pred_uncond, noise_pred_cond = noise_pred.chunk(2)
+                noise_pred = noise_pred_uncond + self.cfg.guidance_scale * (
+                    noise_pred_cond - noise_pred_uncond
+                )
+                # get prev latent
+                latents = self.scheduler.step(noise_pred, t, latents, eta=1)[
+                    "prev_sample"
+                ]
+            latents_final.append(latents)
+
+        latents_final = torch.cat(latents_final)
+        imgs_final = self.decode_latents(latents_final).permute(0, 2, 3, 1)
+
+        return {
+            "imgs_noisy": imgs_noisy,
+            "imgs_1step": imgs_1step,
+            "imgs_final": imgs_final,
+        }
+
+    @torch.cuda.amp.autocast(enabled=False)
+    @torch.no_grad()
+    def validation_step(
+        self,
+        rgb: Float[Tensor, "B H W C"],
+        elevation: Float[Tensor, "B"],
+        azimuth: Float[Tensor, "B"],
+        camera_distances: Float[Tensor, "B"],
+        rgb_as_latents=False,
+        **kwargs,
+    ):
+        rgb_BCHW = rgb.permute(0, 3, 1, 2)
+        latents: Float[Tensor, "B 4 64 64"]
```

Comment:
not sure if this code will still be useful after you make the changes we discussed... but the current version of Zero123 is trained on 256x256x3 inputs converted to 32x32x4 latents... so you might want to make the dimension in pixels here configurable, and default to 256, to minimize VRAM